documentation/design_proposal/002_grafana_v1beta1.md

```diff
@@ -0,0 +1,227 @@
+be opinionated how the default object that we generate should look like just like even more then we do today.
+
+The grafana CRD API is currently a bit sprawling and it can be hard to find how to
+overwrite a number of default values, especially when it comes to deployment configuration.
+
+As a part of the v1beta1 release we want to make this better and at the same time
+give our users more power to customize there deployments.
+
+## Out of scope
+
+In the roadmap for 5.0 there are mentions about moving `dashboardLabelSelector` and `dashboardNamespaceSelector` from the grafana object
+and instead have the grafanadashboard crd to find the grafana it's selected instances.
+This design document don't tale this in to consideration and only focuses on configuration of the grafana instance.
+
+## Verification
+
+- Create integration tests for the new CRD:s
+- Add new e2e tests to cover the new CR.
+
+## Current
+
+Currently the grafana spec looks like this.
+
+```.go
+type GrafanaSpec struct {
+ Config                     GrafanaConfig            `json:"config"`
+ Containers                 []v1.Container           `json:"containers,omitempty"`
+ DashboardLabelSelector     []*metav1.LabelSelector  `json:"dashboardLabelSelector,omitempty"`
+ Ingress                    *GrafanaIngress          `json:"ingress,omitempty"`
+ InitResources              *v1.ResourceRequirements `json:"initResources,omitempty"`
+ Secrets                    []string                 `json:"secrets,omitempty"`
+ ConfigMaps                 []string                 `json:"configMaps,omitempty"`
+ Service                    *GrafanaService          `json:"service,omitempty"`
+ Deployment                 *GrafanaDeployment       `json:"deployment,omitempty"`
+ Resources                  *v1.ResourceRequirements `json:"resources,omitempty"`
+ ServiceAccount             *GrafanaServiceAccount   `json:"serviceAccount,omitempty"`
+ Client                     *GrafanaClient           `json:"client,omitempty"`
+ DashboardNamespaceSelector *metav1.LabelSelector    `json:"dashboardNamespaceSelector,omitempty"`
+ DataStorage                *GrafanaDataStorage      `json:"dataStorage,omitempty"`
+ Jsonnet                    *JsonnetConfig           `json:"jsonnet,omitempty"`
+ BaseImage                  string                   `json:"baseImage,omitempty"`
+ InitImage                  string                   `json:"initImage,omitempty"`
+ LivenessProbeSpec          *LivenessProbeSpec       `json:"livenessProbeSpec,omitempty"`
+ ReadinessProbeSpec         *ReadinessProbeSpec      `json:"readinessProbeSpec,omitempty"`
+}
+```
+
+## Proposal
+
+In short the proposal is about moving the grafana container specific config from the main GrafanaSpec to GrafanaContainer.
+GrafanaContainer should contain [V1.Container](https://pkg.go.dev/k8s.io/api@v0.20.2/core/v1?utm_source=gopls#Container) but with removal of a number of required values like name that will be hard coded to `grafana`.
```

Comment:
I like that idea. But how can we remove fields from the ContainerSpec that we don't want to expose?

Reply:
This was unclear, we need to use something like the operator-tools for the deploymentOverrides but for ContainerSpec.
I don't think it exist in the current package. For starters we can add it to our code and where we only supply the fields that we think should be made available. From the top of my head, the only one we should remove is `name` since it always should be grafana.

Reply:
We can replicate the function logic for deploymentOverride to implement some really simple blocklists, and just reject a CR definition if it tries to override a disallowed field